subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java

```diff
@@ -146,27 +147,35 @@ public ClassSetAnalysisData reduceToTypesAffecting(ClassSetAnalysisData other) {
             }
         }
 
+        Set<String> usedConstantSources = compilerApiData.isSupportsConstantsMapping()
+            ? compilerApiData.getConstantToClassMapping().getConstantDependents().keySet()
+            : classesToConstants.keySet();
+
+        usedClasses.addAll(usedConstantSources);
+
         Map<String, HashCode> classHashes = new HashMap<>(usedClasses.size());
```

Comment:
This datastructure looks a bit inefficient. We create 3 hashmaps from class to whatever. There's confusing redundancy with `ClassAnalysis` actually. We should think about merging those.

Reply:
ClassAnalysis tracks forward dependencies, this here tracks reverse dependencies. Also, in general the keys of these maps can be different. i.e. `dependents` can have more keys than `classHashes` because dependents can contain information about classes outside a JAR that this JAR uses.

Reply:
Yes but hash maps are quite memory intensive and having a domain object which stores the `HashCode` + dependents + constants, even if some of them are empty, is probably more efficient.

Reply:
You would still be left with two maps in that case, because as mentioned above, `dependents.keySet` is a superset of `classHashes.keySet`. So you'd still need to track the dependents of those external classes. The code you see here is a special case where we drop all external classes, because we know it is safe (we're looking at the whole classpath). But that's not how ClassSetAnalysisData generally works.